cosi/service/cosi.go

```diff
@@ -45,64 +46,84 @@ type SignatureResponse struct {
 }
 
 // SignatureRequest treats external request to this service.
-func (cs *CoSi) SignatureRequest(req *SignatureRequest) (network.Message, error) {
-	suite, ok := cs.Suite().(kyber.HashFactory)
-	if !ok {
-		return nil, errors.New("suite is unusable")
+func (s *Service) SignatureRequest(req *SignatureRequest) (network.Message, error) {
+	// generate the tree
+	nNodes := len(req.Roster.List)
+	tree := req.Roster.GenerateNaryTreeWithRoot(nNodes, s.ServerIdentity())
+	if tree == nil {
+		return nil, errors.New("failed to generate tree")
 	}
-
-	if req.Roster.ID.IsNil() {
-		req.Roster.ID = onet.RosterID(uuid.NewV4())
-	}
-
-	_, root := req.Roster.Search(cs.ServerIdentity().ID)
-	if root == nil {
-		return nil, errors.New("Couldn't find a serverIdetity in Roster")
-	}
-	tree := req.Roster.GenerateNaryTreeWithRoot(2, root)
-	tni := cs.NewTreeNodeInstance(tree, tree.Root, cosi.Name)
-	pi, err := cosi.NewProtocol(tni)
+	pi, err := s.CreateProtocol(protocol.DefaultProtocolName, tree)
 	if err != nil {
 		return nil, errors.New("Couldn't make new protocol: " + err.Error())
 	}
-	cs.RegisterProtocolInstance(pi)
-	pcosi := pi.(*cosi.CoSi)
-	pcosi.SigningMessage(req.Message)
-	h := suite.Hash()
-	h.Write(req.Message)
-	response := make(chan []byte)
-	pcosi.RegisterSignatureHook(func(sig []byte) {
-		response <- sig
-	})
+
+	// configure the protocol
+	p := pi.(*protocol.CoSiRootNode)
+	p.CreateProtocol = s.CreateProtocol
+	p.Msg = req.Message
+	// We set NSubtrees to the cube root of n to evenly distribute the load,
+	// i.e. depth (=3) = log_f n, where f is the fan-out (branching factor).
+	p.NSubtrees = int(math.Pow(float64(nNodes), 1.0/3.0))
+	p.Timeout = time.Second * 5
+	if p.NSubtrees < 1 {
+		p.NSubtrees = 1
+	}
+
+	// start the protocol
 	log.Lvl3("Cosi Service starting up root protocol")
-	go pi.Dispatch()
-	go pi.Start()
-	sig := <-response
+	if err = pi.Start(); err != nil {
+		return nil, err
+	}
+
 	if log.DebugVisible() > 1 {
-		fmt.Printf("%s: Signed a message.\n", time.Now().Format("Mon Jan 2 15:04:05 -0700 MST 2006"))
+		log.Printf("%s: Signed a message.\n", time.Now().Format("Mon Jan 2 15:04:05 -0700 MST 2006"))
 	}
-	return &SignatureResponse{
-		Hash:      h.Sum(nil),
-		Signature: sig,
-	}, nil
+
+	// wait for reply
+	var sig []byte
+	select {
+	case sig = <-p.FinalSignature:
+	case <-time.After(p.Timeout + time.Second):
+		return nil, errors.New("protocol timed out")
+	}
+
+	// the hash is the message cosi actually signs, ideally cosi protocol
```

Comment:
Would not be a good time to fix that? If cosi knows it, cosi should give it?

Reply:
The comment is perhaps a bit misleading, I will update it. Cosi knows, but we also need to verify that cosi signed the right message. There are two cases:
1, cosi returns message and signature, so we need to check the message matches the original message and the signature is valid.
2, cosi returns only the signature, then we only check that the signature matches with the original message. If cosi signed a different message, then the final verification will fail.
Hence, in both cases, we arrive at the same result. Whether cosi returns the message is not a big concern in my opinion.